---
title: "Rosalind: A step-by-step Python solution to common bioinformatic problems"
layout: post
date: 2021-01-01 22:10
tag: Bioinformatics; Python
image: https:///David-Boo.github.io/blob/gh-pages/assets/images/Rosalind_Logo.jpg
headerImage: true
projects: true
hidden: true # don't count this post in blog pagination
description: "A step-by-step Python solution to common bioinformatic problems."
category: project
author: David Boo
externalLink: false
---

![Screenshot](https://raw.githubusercontent.com/sergiokopplin/indigo/gh-pages/assets/screen-shot.png)

Example of project - Indigo Minimalist Jekyll Template - [Demo](https://sergiokopplin.github.io/indigo/). This is a simple and minimalist template for Jekyll for those who likes to eat noodles.

| **ID** | **Title**                                                   | **Python solution** |
|:------:|:-----------------------------------------------------------:|:-------------------:|
| DNA    | Counting DNA Nucleotides                                    | 46923               |
| RNA    | Transcribing DNA into RNA                                   | 41858               |
| REVC   | Complementing a Strand of DNA                               | 37925               |
| FIB    | Rabbits and Recurrence Relations                            | 21891               |
| GC     | Computing GC Content                                        | 22042               |
| HAMM   | Counting Point Mutations                                    | 24752               |
| IPRB   | Mendel's First Law                                          | 14627               |
| PROT   | Translating RNA into Protein                                | 19392               |
| SUBS   | Finding a Motif in DNA                                      | 19740               |
| CONS   | Consensus and Profile                                       | 10903               |
| FIBD   | Mortal Fibonacci Rabbits                                    | 9319                |
| IEV    | Calculating Expected Offspring                              | 8424                |
| LIA    | Independent Alleles                                         | 4471                |
| GRPH   | Overlap Graphs                                              | 8898                |
| LCSM   | Finding a Shared Motif                                      | 7681                |
| MPRT   | Finding a Protein Motif                                     | 4758                |
| MRNA   | Inferring mRNA from Protein                                 | 7273                |
| ORF    | Open Reading Frames                                         | 5636                |
| PERM   | Enumerating Gene Orders                                     | 9856                |
| PRTM   | Calculating Protein Mass                                    | 9424                |
| REVP   | Locating Restriction Sites                                  | 5956                |
| SPLC   | RNA Splicing                                                | 6634                |
| PROB   | Introduction to Random Strings                              | 3596                |
| LEXF   | Enumerating k-mers Lexicographically                        | 5512                |
| LGIS   | Longest Increasing Subsequence                              | 2433                |
| LONG   | Genome Assembly as Shortest Superstring                     | 2716                |
| PMCH   | Perfect Matchings and RNA Secondary Structures              | 2654                |
| PPER   | Partial Permutations                                        | 3618                |
| SIGN   | Enumerating Oriented Gene Orderings                         | 3607                |
| SSEQ   | Finding a Spliced Motif                                     | 3957                |
| TRAN   | Transitions and Transversions                               | 3861                |
| TREE   | Completing a Tree                                           | 3195                |
| CAT    | Catalan Numbers and RNA Secondary Structures                | 1088                |
| CORR   | Error Correction in Reads                                   | 1723                |
| INOD   | Counting Phylogenetic Ancestors                             | 2353                |
| KMER   | k-Mer Composition                                           | 2603                |
| KMP    | Speeding Up Motif Finding                                   | 2073                |
| LCSQ   | Finding a Shared Spliced Motif                              | 1812                |
| LEXV   | Ordering Strings of Varying Length Lexicographically        | 2911                |
| MMCH   | Maximum Matchings and RNA Secondary Structures              | 1313                |
| PDST   | Creating a Distance Matrix                                  | 1911                |
| REAR   | Reversal Distance                                           | 907                 |
| RSTR   | Matching Random Motifs                                      | 1456                |
| SSET   | Counting Subsets                                            | 2183                |
| ASPC   | Introduction to Alternative Splicing                        | 1384                |
| EDIT   | Edit Distance                                               | 1302                |
| EVAL   | Expected Number of Restriction Sites                        | 1090                |
| MOTZ   | Motzkin Numbers and RNA Secondary Structures                | 678                 |
| NWCK   | Distances in Trees                                          | 834                 |
| SCSP   | Interleaving Two Motifs                                     | 808                 |
| SETO   | Introduction to Set Operations                              | 1705                |
| SORT   | Sorting by Reversals                                        | 684                 |
| SPEC   | Inferring Protein from Spectrum                             | 1322                |
| TRIE   | Introduction to Pattern Matching                            | 1029                |
| CONV   | Comparing Spectra with the Spectral Convolution             | 843                 |
| CTBL   | Creating a Character Table                                  | 475                 |
| DBRU   | Constructing a De Bruijn Graph                              | 868                 |
| EDTA   | Edit Distance Alignment                                     | 860                 |
| FULL   | Inferring Peptide from Full Spectrum                        | 602                 |
| INDC   | Independent Segregation of Chromosomes                      | 689                 |
| ITWV   | Finding Disjoint Motifs in a Gene                           | 336                 |
| LREP   | Finding the Longest Multiple Repeat                         | 465                 |
| NKEW   | Newick Format with Edge Weights                             | 522                 |
| RNAS   | Wobble Bonding and RNA Secondary Structures                 | 471                 |
| AFRQ   | Counting Disease Carriers                                   | 585                 |
| CSTR   | Creating a Character Table from Genetic Strings             | 310                 |
| CTEA   | Counting Optimal Alignments                                 | 331                 |
| CUNR   | Counting Unrooted Binary Trees                              | 299                 |
| GLOB   | Global Alignment with Scoring Matrix                        | 619                 |
| PCOV   | Genome Assembly with Perfect Coverage                       | 627                 |
| PRSM   | Matching a Spectrum to a Protein                            | 452                 |
| QRT    | Quartets                                                    | 248                 |
| SGRA   | Using the Spectrum Graph to Infer Peptides                  | 393                 |
| SUFF   | Encoding Suffix Trees                                       | 337                 |
| CHBP   | Character-Based Phylogeny                                   | 165                 |
| CNTQ   | Counting Quartets                                           | 176                 |
| EUBT   | Enumerating Unrooted Binary Trees                           | 162                 |
| GASM   | Genome Assembly Using Reads                                 | 347                 |
| GCON   | Global Alignment with Constant Gap Penalty                  | 382                 |
| LING   | Linguistic Complexity of a Genome                           | 204                 |
| LOCA   | Local Alignment with Scoring Matrix                         | 410                 |
| MEND   | Inferring Genotype from a Pedigree                          | 267                 |
| MGAP   | Maximizing the Gap Symbols of an Optimal Alignment          | 216                 |
| MREP   | Identifying Maximal Repeats                                 | 184                 |
| MULT   | Multiple Alignment                                          | 221                 |
| PDPL   | Creating a Restriction Map                                  | 239                 |
| ROOT   | Counting Rooted Binary Trees                                | 248                 |
| SEXL   | Sex-Linked Inheritance                                      | 446                 |
| SPTD   | Phylogeny Comparison with Split Distance                    | 177                 |
| WFMD   | The Wright-Fisher Model of Genetic Drift                    | 349                 |
| ALPH   | Alignment-Based Phylogeny                                   | 125                 |
| ASMQ   | Assessing Assembly Quality with N50 and N75                 | 275                 |
| CSET   | Fixing an Inconsistent Character Set                        | 135                 |
| EBIN   | Wright-Fisher's Expected Behavior                           | 291                 |
| FOUN   | The Founder Effect and Genetic Drift                        | 276                 |
| GAFF   | Global Alignment with Scoring Matrix and Affine Gap Penalty | 314                 |
| GREP   | Genome Assembly with Perfect Coverage and Repeats           | 210                 |
| OAP    | Overlap Alignment                                           | 189                 |
| QRTD   | Quartet Distance                                            | 88                  |
| SIMS   | Finding a Motif with Modifications                          | 225                 |
| SMGB   | Semiglobal Alignment                                        | 191                 |
| KSIM   | Finding All Similar Motifs                                  | 84                  |
| LAFF   | Local Alignment with Affine Gap Penalty                     | 186                 |
| OSYM   | Isolating Symbols in Alignments                             | 138                 |
| RSUB   | Identifying Reversing Substitutions                         | 103                 |

---

What has inside?

- Gulp
- BrowserSync
- Stylus
- SVG
- Travis
- No JS
- [98/100](https://developers.google.com/speed/pagespeed/insights/?url=http%3A%2F%2Fsergiokopplin.github.io%2Findigo%2F)

---

[Check it out](https://sergiokopplin.github.io/indigo/) here.
If you need some help, just [tell me](https://github.com/sergiokopplin/indigo/issues).
