---
title: "Rosalind: A step-by-step Python solution to common bioinformatic problems"
layout: post
date: 2021-01-03 22:10
tag:
- Bioinformatics
- Python
image: /assets/images/Rosalind_Logo.jpg
headerImage: true
projects: true
hidden: true # don't count this post in blog pagination
description: "A step-by-step Python solution to common bioinformatic problems."
category: project
author: davidboo
externalLink: false
---

[Rosalind Franklin](https://www.rfi.ac.uk/about/rosalind-franklin/) was an outstanding, successful researcher in several fields of knowledge. Specifically, she made great contributions to the understanding of the fine molecular structures of DNA, RNA, viruses, coal and graphite. 

<p align="center">
  <img src="https://github.com/David-boo/David-Boo.github.io/blob/gh-pages/assets/images/Rosalind_Franklin.jpg?raw=true">
</p>


Franklin is best known for her work on the [X-ray diffraction images of DNA](https://www.rosalindfranklinsociety.org/rosalind-franklin) which led to discovery of DNA double helix. Her X-ray diffraction images confirming the helical structure of DNA were shown to Watson without her approval or knowledge, and gave birth to Crick and Watson's 1953 hypothesis regarding the structure of DNA. Her work was published third, in the series of three DNA Nature articles, led by the paper of Watson and Crick which only hinted at her contribution to their hypothesis. 

Rosalind Franklin is often associated only to this infamous event, and that is really unfair considering how great of a scientific (and role model) she was. The [Rosalind Project](http://rosalind.info/about/) is just a small tribute to her importance. <span class="evidence">**Rosalind is a platform for learning bioinformatics through computational problems of varying difficulty that are extracted from real challenges of molecular biology**.</span>

Created by the University of California at San Diego and Saint Petersburg Academic University, and inspired on the [Project Euler](https://david-boo.github.io//project-euler-solutions/), Rosalind try inspire a new generation of bioinformatics students to develop vital programming skills at their own pace in a unique environment. However, most of the times students without a coding background (such as biology-focused students) tend to be afraid and end up not learning and developing new programming skills.

In this page you will find my personal solutions to the problems, using only Python. They have been part of my process trying to learn Python and so, code can be messy or not optimized (sorry in advance). <span class="evidence">Every solution is not just plain code, as **I've written small explanations on every line, so it can serve as a Python learning experience for other bio-focused studentes, just as it was for me**.</span> Also, I've tried to include simple, brief explanations of most common biology terms, aimed at students that may be lacking in this area (IT background students).

**I strongly encourage you to try and solve the problems without looking at any solution**. Even if at first sight a problem feels overwhelming, give it a try and come back a few days later. You will be surprised at all the things you are capable of doing and you will learn even more. 

And just remember, if everything goes wrong, here you'll find some help. :)

---


|ID|Title|Python Solution|
|:-----|:-----:|-----:|
|DNA|[Counting DNA Nucleotides](http://rosalind.info/problems/dna/)|[Rosalind_DNA](https://github.com/David-boo/Rosalind/blob/main/Rosalind_DNA.py)|
|RNA|[Transcribing DNA into RNA](http://rosalind.info/problems/rna/)|[Rosalind_RNA](https://github.com/David-boo/Rosalind/blob/main/Rosalind_RNA.py)|
|REVC|[Complementing a Strand of DNA](http://rosalind.info/problems/revc/)|[Rosalind_REVC](https://github.com/David-boo/Rosalind/blob/main/Rosalind_REVC.py)|
|FIB|[Rabbits and Recurrence Relations](http://rosalind.info/problems/fib/)|[Rosalind_FIB](https://github.com/David-boo/Rosalind/blob/main/Rosalind_FIB.py)|
|GC|[Computing GC Content](http://rosalind.info/problems/gc/)|[Rosalind_GC](https://github.com/David-boo/Rosalind/blob/main/Rosalind_GC.py)|
|HAMM|[Counting Point Mutations](http://rosalind.info/problems/hamm/)|[Rosalind_HAMM](https://github.com/David-boo/Rosalind/blob/main/Rosalind_HAMM.py)|
|IPRB|[Mendel's First Law](http://rosalind.info/problems/iprb/)|[Rosalind_IPRB](https://github.com/David-boo/Rosalind/blob/main/Rosalind_IPRB.py)|
|PROT|[Translating RNA into Protein](http://rosalind.info/problems/prot/)|[Rosalind_PROT](https://github.com/David-boo/Rosalind/blob/main/Rosalind_PROT.py)|
|SUBS|[Finding a Motif in DNA](http://rosalind.info/problems/subs/)|[Rosalind_SUBS](https://github.com/David-boo/Rosalind/blob/main/Rosalind_SUBS.py)|
|CONS|[Consensus and Profile](http://rosalind.info/problems/cons/)|[Rosalind_CONS](https://github.com/David-boo/Rosalind/blob/main/Rosalind_CONS.py)|
|FIBD|[Mortal Fibonacci Rabbits](http://rosalind.info/problems/fibd/)|[Rosalind_FIBD](https://github.com/David-boo/Rosalind/blob/main/Rosalind_FIBD.py)|
|IEV|[Calculating Expected Offspring](http://rosalind.info/problems/iev/)|[Rosalind_IEV](https://github.com/David-boo/Rosalind/blob/main/Rosalind_IEV.py)|
|LIA|[Independent Alleles](http://rosalind.info/problems/lia/)|[Rosalind_LIA](https://github.com/David-boo/Rosalind/blob/main/Rosalind_LIA.py)|
|GRPH|[Overlap Graphs](http://rosalind.info/problems/grph/)|[8898](http://rosalind.info/problems/grph/recent/)|
|LCSM|[Finding a Shared Motif](http://rosalind.info/problems/lcsm/)||
|MPRT|[Finding a Protein Motif](http://rosalind.info/problems/mprt/)||
|MRNA|[Inferring mRNA from Protein](http://rosalind.info/problems/mrna/)||
|ORF|[Open Reading Frames](http://rosalind.info/problems/orf/)||
|PERM|[Enumerating Gene Orders](http://rosalind.info/problems/perm/)||
|PRTM|[Calculating Protein Mass](http://rosalind.info/problems/prtm/)||
|REVP|[Locating Restriction Sites](http://rosalind.info/problems/revp/)||
|SPLC|[RNA Splicing](http://rosalind.info/problems/splc/)||
|PROB|[Introduction to Random Strings](http://rosalind.info/problems/prob/)||
|LEXF|[Enumerating k-mers Lexicographically](http://rosalind.info/problems/lexf/)||
|LGIS|[Longest Increasing Subsequence](http://rosalind.info/problems/lgis/)||
|LONG|[Genome Assembly as Shortest Superstring](http://rosalind.info/problems/long/)||
|PMCH|[Perfect Matchings and RNA Secondary Structures](http://rosalind.info/problems/pmch/)||
|PPER|[Partial Permutations](http://rosalind.info/problems/pper/)||
|SIGN|[Enumerating Oriented Gene Orderings](http://rosalind.info/problems/sign/)||
|SSEQ|[Finding a Spliced Motif](http://rosalind.info/problems/sseq/)||
|TRAN|[Transitions and Transversions](http://rosalind.info/problems/tran/)||
|TREE|[Completing a Tree](http://rosalind.info/problems/tree/)||
|CAT|[Catalan Numbers and RNA Secondary Structures](http://rosalind.info/problems/cat/)||
|CORR|[Error Correction in Reads](http://rosalind.info/problems/corr/)||
|INOD|[Counting Phylogenetic Ancestors](http://rosalind.info/problems/inod/)||
|KMER|[k-Mer Composition](http://rosalind.info/problems/kmer/)||
|KMP|[Speeding Up Motif Finding](http://rosalind.info/problems/kmp/)||
|LCSQ|[Finding a Shared Spliced Motif](http://rosalind.info/problems/lcsq/)||
|LEXV|[Ordering Strings of Varying Length Lexicographically](http://rosalind.info/problems/lexv/)||
|MMCH|[Maximum Matchings and RNA Secondary Structures](http://rosalind.info/problems/mmch/)||
|PDST|[Creating a Distance Matrix](http://rosalind.info/problems/pdst/)||
|REAR|[Reversal Distance](http://rosalind.info/problems/rear/)||
|RSTR|[Matching Random Motifs](http://rosalind.info/problems/rstr/)||
|SSET|[Counting Subsets](http://rosalind.info/problems/sset/)||
|ASPC|[Introduction to Alternative Splicing](http://rosalind.info/problems/aspc/)||
|EDIT|[Edit Distance](http://rosalind.info/problems/edit/)||
|EVAL|[Expected Number of Restriction Sites](http://rosalind.info/problems/eval/)||
|MOTZ|[Motzkin Numbers and RNA Secondary Structures](http://rosalind.info/problems/motz/)||
|NWCK|[Distances in Trees](http://rosalind.info/problems/nwck/)||
|SCSP|[Interleaving Two Motifs](http://rosalind.info/problems/scsp/)||
|SETO|[Introduction to Set Operations](http://rosalind.info/problems/seto/)||
|SORT|[Sorting by Reversals](http://rosalind.info/problems/sort/)||
|SPEC|[Inferring Protein from Spectrum](http://rosalind.info/problems/spec/)||
|TRIE|[Introduction to Pattern Matching](http://rosalind.info/problems/trie/)||
|CONV|[Comparing Spectra with the Spectral Convolution](http://rosalind.info/problems/conv/)||
|CTBL|[Creating a Character Table](http://rosalind.info/problems/ctbl/)||
|DBRU|[Constructing a De Bruijn Graph](http://rosalind.info/problems/dbru/)||
|EDTA|[Edit Distance Alignment](http://rosalind.info/problems/edta/)||
|FULL|[Inferring Peptide from Full Spectrum](http://rosalind.info/problems/full/)||
|INDC|[Independent Segregation of Chromosomes](http://rosalind.info/problems/indc/)||
|ITWV|[Finding Disjoint Motifs in a Gene](http://rosalind.info/problems/itwv/)||
|LREP|[Finding the Longest Multiple Repeat](http://rosalind.info/problems/lrep/)||
|NKEW|[Newick Format with Edge Weights](http://rosalind.info/problems/nkew/)||
|RNAS|[Wobble Bonding and RNA Secondary Structures](http://rosalind.info/problems/rnas/)||
|AFRQ|[Counting Disease Carriers](http://rosalind.info/problems/afrq/)||
|CSTR|[Creating a Character Table from Genetic Strings](http://rosalind.info/problems/cstr/)||
|CTEA|[Counting Optimal Alignments](http://rosalind.info/problems/ctea/)||
|CUNR|[Counting Unrooted Binary Trees](http://rosalind.info/problems/cunr/)||
|GLOB|[Global Alignment with Scoring Matrix](http://rosalind.info/problems/glob/)||
|PCOV|[Genome Assembly with Perfect Coverage](http://rosalind.info/problems/pcov/)||
|PRSM|[Matching a Spectrum to a Protein](http://rosalind.info/problems/prsm/)||
|QRT|[Quartets](http://rosalind.info/problems/qrt/)||
|SGRA|[Using the Spectrum Graph to Infer Peptides](http://rosalind.info/problems/sgra/)||
|SUFF|[Encoding Suffix Trees](http://rosalind.info/problems/suff/)||
|CHBP|[Character-Based Phylogeny](http://rosalind.info/problems/chbp/)||
|CNTQ|[Counting Quartets](http://rosalind.info/problems/cntq/)||
|EUBT|[Enumerating Unrooted Binary Trees](http://rosalind.info/problems/eubt/)||
|GASM|[Genome Assembly Using Reads](http://rosalind.info/problems/gasm/)||
|GCON|[Global Alignment with Constant Gap Penalty](http://rosalind.info/problems/gcon/)||
|LING|[Linguistic Complexity of a Genome](http://rosalind.info/problems/ling/)||
|LOCA|[Local Alignment with Scoring Matrix](http://rosalind.info/problems/loca/)||
|MEND|[Inferring Genotype from a Pedigree](http://rosalind.info/problems/mend/)||
|MGAP|[Maximizing the Gap Symbols of an Optimal Alignment](http://rosalind.info/problems/mgap/)||
|MREP|[Identifying Maximal Repeats](http://rosalind.info/problems/mrep/)||
|MULT|[Multiple Alignment](http://rosalind.info/problems/mult/)||
|PDPL|[Creating a Restriction Map](http://rosalind.info/problems/pdpl/)||
|ROOT|[Counting Rooted Binary Trees](http://rosalind.info/problems/root/)||
|SEXL|[Sex-Linked Inheritance](http://rosalind.info/problems/sexl/)||
|SPTD|[Phylogeny Comparison with Split Distance](http://rosalind.info/problems/sptd/)||
|WFMD|[The Wright-Fisher Model of Genetic Drift](http://rosalind.info/problems/wfmd/)||
|ALPH|[Alignment-Based Phylogeny](http://rosalind.info/problems/alph/)||
|ASMQ|[Assessing Assembly Quality with N50 and N75](http://rosalind.info/problems/asmq/)||
|CSET|[Fixing an Inconsistent Character Set](http://rosalind.info/problems/cset/)||
|EBIN|[Wright-Fisher's Expected Behavior](http://rosalind.info/problems/ebin/)||
|FOUN|[The Founder Effect and Genetic Drift](http://rosalind.info/problems/foun/)||
|GAFF|[Global Alignment with Scoring Matrix and Affine Gap Penalty](http://rosalind.info/problems/gaff/)||
|GREP|[Genome Assembly with Perfect Coverage and Repeats](http://rosalind.info/problems/grep/)||
|OAP|[Overlap Alignment](http://rosalind.info/problems/oap/)||
|QRTD|[Quartet Distance](http://rosalind.info/problems/qrtd/)||
|SIMS|[Finding a Motif with Modifications](http://rosalind.info/problems/sims/)||
|SMGB|[Semiglobal Alignment](http://rosalind.info/problems/smgb/)||
|KSIM|[Finding All Similar Motifs](http://rosalind.info/problems/ksim/)||
|LAFF|[Local Alignment with Affine Gap Penalty](http://rosalind.info/problems/laff/)||
|OSYM|[Isolating Symbols in Alignments](http://rosalind.info/problems/osym/)||
|RSUB|[Identifying Reversing Substitutions](http://rosalind.info/problems/rsub/)||


---

Thank you for your time and attention. Feel free to [reach out](https://david-boo.github.io/) if you have any questions or inquiries. I'd love to hear from you!

